src/cpp/runner/main.cpp

```diff
@@ -9,26 +10,144 @@
+        const auto current_run_time = duration_cast<duration<double>>(t2 - t1).count();
+        if (current_run_time > minimum_measurable_time)
+        {
+            samples.push_back(current_run_time / repeats);
+            total_time += current_run_time;
+            break;
+        }
+    }
+
+    for (auto run = 1; (run < nruns) && (total_time < time_limit); run++)
+    {
+        auto t1 = high_resolution_clock::now();
+        call_member_function(test, func, repeats);
+        auto t2 = high_resolution_clock::now();
+        const auto current_run_time = duration_cast<duration<double>>(t2 - t1).count();
+        samples.push_back(current_run_time / repeats);
+        total_time += current_run_time;
+    }
+
+    const auto min_sample = *std::min_element(std::begin(samples), std::end(samples));
+
+    return min_sample;
+}
+
+std::string filepath_to_basename(const std::string& filepath)
+{
+    const auto filename = filepath.substr(filepath.find_last_of("/\\") + 1);
+    const auto dot = filename.find_last_of('.');
+    auto basename = filename.substr(0, dot);
+
+    return basename;
+}
 
-}*/
+void save_time_to_file(const string& filepath, const double objective_time, const double derivative_time)
+{
+    std::ofstream out(filepath);
+    out << std::scientific << objective_time << "\t" << derivative_time;
+    out.close();
+}
+
+void save_objective_to_file(const string& filepath, const double value)
+{
+    std::ofstream out(filepath);
+    out << std::scientific << value;
+    out.close();
+}
+
+void save_gradient_to_file(const string& filepath, int gradient_size, double* gradient)
+{
+    std::ofstream out(filepath);
+
+    for (auto i = 0; i < gradient_size; i++)
+    {
+        out << std::scientific << gradient[i] << "\t";
+    }
+
+    out.close();
+}
 
-int main(int argc, const char* argv[])
+int main(const int argc, const char* argv[])
 {
-	if (argc < 7) {
-		std::cerr << "usage: CPPRunner modulePath dir_in dir_out file_basename nruns_F nruns_J [-rep]\n";
-		return 1;
-	}
+    try {
+        if (argc < 8) {
+            std::cerr << "usage: CPPRunner module_path input_filepath output_dir minimum_measurable_time nruns_F nruns_J time_limit [-rep]\n";
+            return 1;
+        }
+
+        const auto module_path = argv[1];
+        const string input_filepath(argv[2]);
+        const string output_dir(argv[3]);
+        const auto minimum_measurable_time = std::stod(argv[4]);
+        const auto nruns_F = std::stoi(argv[5]);
+        const auto nruns_J = std::stoi(argv[6]);
+        const auto time_limit = std::stod(argv[7]);
+
+        // read only 1 point and replicate it?
+        const auto replicate_point = (argc > 8 && string(argv[8]) == "-rep");
+
+        ModuleLoader module_loader(module_path);
+        auto test = module_loader.GetTest();
+
+        auto inputs = read_input_data(input_filepath, replicate_point);
+        auto gradient_size = (inputs.k * (inputs.d + 1) * (inputs.d + 2)) / 2;
```

Comment:
This size is a part of the output vector. No need to compute it separately here.